### PR TITLE
feat: add memory-efficient DoRA kernel via use_dora="kernel"

### PR DESCRIPTION
**Pull request by peft-jambot**
## Summary

This PR adds a memory-efficient DoRA implementation that avoids materializing the full `lora_B @ lora_A` matrix, enabled via `LoraConfig(use_dora="kernel")`.

## Background

Standard DoRA computes the column-wise weight norm `||W + s*BA||_c` by first materializing the full `out × in` matrix `BA = B @ A`. For large models (e.g. `out=4096, in=11008`), this matmul is expensive in both memory and runtime.

## Approach

The kernel decomposes the squared norm using the low-rank structure of `BA = B @ A`:

```
||W + s*BA||² = ||W||² + 2*s*<W, BA> + s²*||BA||²
```

where:
- `<W, BA> = (W @ Aᵀ ⊙ B).sum(dim=1)` — only `out × r` intermediate
- `||BA||² = (B @ (A@Aᵀ) ⊙ B).sum(dim=1)` — only `r × r` then `out × r` intermediate

This reduces peak memory from O(out*in) to O(out*r + r²).

## Benchmark results (CPU)

| Dimensions | Rank | Standard (ms) | Kernel (ms) | Speedup |
|---|---|---|---|---|
| 4096 × 4096 | 8 | 255 | 54 | **4.76x** |
| 11008 × 4096 | 64 | 940 | 163 | **5.80x** |

The speedup increases with larger dimensions because the standard `B@A` matmul is O(out × in × r) while the kernel intermediates are O(out × in × r + out × r²), which is significantly cheaper when r << min(out, in).

## Implementation

- `src/peft/tuners/lora/dora_kernel.py`: `DoraKernelLinearLayer` — the kernel layer with low-rank weight norm computation
- `src/peft/tuners/lora/variants.py`: `DoraKernelLinearVariant` — variant dispatch, reuses standard DoRA merge/unmerge
- `src/peft/tuners/lora/config.py`: `use_dora` field changed from `bool` to `Union[bool, Literal["kernel"]]`
- `src/peft/tuners/lora/layer.py`: `resolve_lora_variant` dispatches to kernel variant when `use_dora="kernel"`
- `scripts/benchmark_dora_kernel.py`: benchmark script for comparing standard vs kernel

## Limitations (first version)

- Only `nn.Linear` base layers are supported (no conv / embedding — they fall back to standard DoRA)
- When `lora_dropout > 0` in training mode, the kernel falls back to computing the full `BA` matrix (same as standard DoRA)
- When `base_result` is not provided (training with dropout), the full path is used

## Tests

- `test_dora_kernel_matches_standard` (parametrized by bias/no-bias, dropout/no-dropout): verifies forward outputs and gradients match standard DoRA
- `test_dora_kernel_merge_unmerge`: verifies merge and unmerge produce correct results
- `test_dora_kernel_save_load`: verifies save/load round-trip preserves outputs

All tests pass:
```
pytest tests/test_custom_models.py -k "test_dora_kernel" -v
# 6 passed
```

Existing DoRA tests (578 tests) also pass:
```
pytest tests/test_custom_models.py -k "dora and not dismiss and not hira and not cran"
# 578 passed, 31 skipped
```

## Related issue

This is a research task from [#7 on peft-jambot/peft](https://github.com/peft-jambot/peft/issues/7) — "Investigate if DoRA could be implemented more efficiently with a specialized kernel."

## AI assistance

This PR was created with AI assistance. The implementation, tests, and benchmark were written and verified by the AI agent.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks for the PR. Since the kernel DoRA version doesn't add any extra dependencies (say, triton), it could theoretically replace the existing DoRA implementation instead of being an alternative option.

I ran a couple of tests with this new DoRA kernel on the MetaMath and image generation benchmarks. There is no TODO following from this, just documenting my findings:

## MetaMathQA

250 steps total.

### normal DoRA:

```
{"step": "  250", "samples": "   1000", "lr": "5.87e-05", "loss avg": "0.8182", "valid acc": "0.380", "gen valid tokens": 8812, "train time": "34.5s", "eval time": "8.2s", "train tokens / sec": "6134", "mem allocated": "6925143398", "mem reserved": "15535240643", "elapsed time": "1min 5s"}
{"step": "  500", "samples": "   2000", "lr": "0.00e+00", "loss avg": "0.6911", "valid acc": "0.360", "gen valid tokens": 9322, "train time": "33.8s", "eval time": "9.8s", "train tokens / sec": "6146", "mem allocated": "6917772009", "mem reserved": "15111674659", "elapsed time": "1min 52s"}
```

```
Test accuracy: 0.372
accelerator memory max: 20548MB
accelerator memory reserved avg: 14613MB
accelerator memory reserved 99th percentile: 20548MB
train time: 446.7330271039682s
total time: 468.53s
```

### DoRA kernel:

```
{"step": "  250", "samples": "   1000", "lr": "5.87e-05", "loss avg": "0.8182", "valid acc": "0.420", "gen valid tokens": 9004, "train time": "31.9s", "eval time": "10.3s", "train tokens / sec": "6630", "mem allocated": "6927350024", "mem reserved": "15417070322", "elapsed time": "1min 4s"}
{"step": "  500", "samples": "   2000", "lr": "0.00e+00", "loss avg": "0.6911", "valid acc": "0.480", "gen valid tokens": 9488, "train time": "31.3s", "eval time": "9.6s", "train tokens / sec": "6653", "mem allocated": "6919892146", "mem reserved": "14996490682", "elapsed time": "1min 48s"}
```

```
Test accuracy: 0.368
accelerator memory max: 20022MB
accelerator memory reserved avg: 14502MB
accelerator memory reserved 99th percentile: 20022MB
train time: 376.729352149996s
total time: 400.51s
file size of checkpoint: 35.5MB
```

## image gen

150 steps total

### normal DoRA

`{"step": " 100", "samples": "  200", "lr": "1.00e-04", "loss avg": "0.7733", "valid sim": "0.6679", "train time": "126.4s", "eval time": "26.6s", "mem allocated": "8905764997", "mem reserved": "12381564436", "elapsed time": "2min 38s"}`

```
Test DINOv2 similarity: 0.6188
Test drift:             0.2541
Saved PEFT checkpoint to /tmp/tmp2bzqt4t8
Experiment run was categorized as a test run on branch enh-better-dora
accelerator memory max: 11880MB
accelerator memory reserved avg: 11831MB
accelerator memory reserved 99th percentile: 11880MB
train time: 480.60444636497414s
total time: 516.53s
file size of checkpoint: 149.7MB
```

### DoRA kernel:

`{"step": " 100", "samples": "  200", "lr": "1.00e-04", "loss avg": "0.7736", "valid sim": "0.6295", "train time": "112.6s", "eval time": "21.2s", "mem allocated": "8907352279", "mem reserved": "11714670100", "elapsed time": "2min 19s"}`

```
Test DINOv2 similarity: 0.6585
Test drift:             0.2502
Saved PEFT checkpoint to /tmp/tmpmldqvod5
Experiment run was categorized as a test run on branch enh-better-dora
accelerator memory max: 11304MB
accelerator memory reserved avg: 11215MB
accelerator memory reserved 99th percentile: 11304MB
train time: 400.2186048569856s
total time: 432.66s
file size of checkpoint: 149.7MB
```

In conclusion, the train loss tracks very closely, which is another confirmation that the two approaches are equivalent. In these more realistic settings, the difference between kernel vs no kernel is, however, quite small. It's probably not worth the added complexity.
